Callout styling

shubham7109 opened the pull request:
Related to issue: https://devtopia.esri.com/runtime/kotlin/issues/3937

### Description:
PR to add the `CalloutTheme.kt` and `DefaultThemeTokens.kt` to enable custom styling on the Callout. 

<details><summary>Test Code: </summary>

Screenshot: 
![Screenshot 2024-07-09 at 3 57 43 PM](https://github.com/Esri/arcgis-maps-sdk-kotlin-toolkit/assets/14892474/059e7ad4-eaee-4262-8f33-951a9bf0892b)

```kotlin

import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.DpSize
import androidx.compose.ui.unit.dp
import com.arcgismaps.toolkit.geoviewcompose.MapView
import com.arcgismaps.toolkit.geoviewcompose.theme.CalloutDefaults

@Composable
fun GraphicScreen(mapViewModel: MapViewModel) {

    val tappedPoint = mapViewModel.mapPoint.collectAsState().value
    val calloutTheme = CalloutDefaults.colors(
        backgroundColor = Color.LightGray,
        borderColor = Color.Red
    )

    val calloutShape = CalloutDefaults.shapes(
        cornerRadius = 0.dp,
        leaderSize = DpSize(25.dp, 20.dp),
        borderWidth = 5.dp,
        calloutContentPadding = PaddingValues(25.dp),
        minSize = DpSize(100.dp, 100.dp)
    )

    Scaffold { contentPadding ->
        MapView(
            modifier = Modifier.fillMaxSize().padding(contentPadding),
            arcGISMap = mapViewModel.arcGISMap,
            graphicsOverlays = remember { listOf(mapViewModel.tapLocationGraphicsOverlay) },
            onSingleTapConfirmed = mapViewModel::setMapPoint
        ) {
            if (tappedPoint != null) {
                Callout(
                    location = tappedPoint,
                    colorScheme = calloutTheme,
                    shape = calloutShape
                ) {
                    Text(text = "HELLO WORLD")
                }
            }
        }
    }
}
```

</details>


### Pre-merge Checklist

<!-- Tick one box for each section -->
- a [vTest](https://runtime-kotlin.esri.com/view/all/job/vtest/job/toolkit/) Job for this PR has been run
  - [ ] link:
- Unit and/or integration tests have been added to exercise this PR's logic, and the tests are passing:
  - [ ] Yes
  - [ ] No

shubham7109 commented on microapps/MapViewCalloutApp/app/src/main/java/com/arcgismaps/toolkit/mapviewcalloutapp/screens/TapLocationScreen.kt:
```diff
@@ -133,7 +135,8 @@ fun TapLocationScreen(viewModel: MapViewModel) {
                                             "<i>wkid</i> = ${it.spatialReference?.wkid}",
-                                    htmlFlag = HtmlCompat.FROM_HTML_MODE_COMPACT
+                                    htmlFlag = HtmlCompat.FROM_HTML_MODE_COMPACT,
+                                    textColor = MaterialTheme.colorScheme.onBackground,
```
I had to update the microapp to be able to update the text color, since the HtmlText here is wrapped with an AndroidView.

shubham7109 replied:
Correct!

shubham7109 commented on toolkit/geoview-compose/src/main/java/com/arcgismaps/toolkit/geoviewcompose/GeoViewScope.kt:
```diff
@@ -278,22 +286,24 @@ public sealed class GeoViewScope protected constructor(private val geoView: GeoV
-                )
-                {
-                    content.invoke(this)
+                CalloutTheme{
```
Is this really needed? I am able to see the MicroApp theme being applied with/without this `CalloutTheme`

shubham7109 commented on toolkit/geoview-compose/src/main/java/com/arcgismaps/toolkit/geoviewcompose/theme/CalloutTheme.kt:
```diff
@@ -0,0 +1,165 @@
+
+     * @since 200.5.0
+     */
+    @Composable
+    public fun colorScheme(
+        calloutColorsProperties: CalloutColorsProperties = colorProperties(),
+        calloutShapeProperties: CalloutShapeProperties = shapeProperties()
+    ): CalloutColorScheme {
+        return CalloutColorScheme(
+            calloutColorsProperties = calloutColorsProperties,
+            calloutShapeProperties = calloutShapeProperties
+        )
+    }
```
@sorenoid @kaushikrw 

I might reconsider this `colorScheme` to be independent of the Shape properties, I think it might be bit misleading for users to modify the Callout shape from the ColorScheme. 

Looking at the Compose's Button [source code](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/material3/material3/src/commonMain/kotlin/androidx/compose/material3/Button.kt;l=112?q=file:androidx%2Fcompose%2Fmaterial3%2FButton.kt%20function:Button), they have a separate parameter for the shape and color. We can do something similar with Callout function parameters:

```kotlin
    @Composable
    public fun Callout(
        location: Point,
        modifier: Modifier = Modifier,
        offset: Offset = Offset.Zero,
        rotateOffsetWithGeoView: Boolean = false,
        colorScheme: CalloutColorScheme = CalloutDefaults.colorScheme(),
        shape: CalloutShape = CalloutDefaults.shape(),
        content: @Composable BoxScope.() -> Unit
    ) { ....
```

Would like any feedback if there would be concerns with this approach.

sorenoid replied:
although I would call it `CalloutShapes` rather than `CalloutShape`, adnd `CalloutDefaults.shapes()

shubham7109 commented on toolkit/geoview-compose/src/main/java/com/arcgismaps/toolkit/geoviewcompose/GeoViewScope.kt:
```diff
@@ -278,22 +286,24 @@ public sealed class GeoViewScope protected constructor(private val geoView: GeoV
+                                leaderWidth = with(LocalDensity.current) { colorScheme.calloutShapeProperties.leaderSize.width.toPx() },
+                                leaderHeight = with(LocalDensity.current) { colorScheme.calloutShapeProperties.leaderSize.height.toPx() },
+                                minSize = colorScheme.calloutShapeProperties.minSize
+                            )
```
Wrapped the callout Box to here with `with(LocalDensity..` (https://github.com/Esri/arcgis-maps-sdk-kotlin-toolkit/pull/503/commits/90c36532c7c12eacf846422db30076db161b434d)

shubham7109 commented on toolkit/geoview-compose/src/main/java/com/arcgismaps/toolkit/geoviewcompose/theme/CalloutTheme.kt:
```diff
@@ -0,0 +1,165 @@
+)
+
+@Immutable
+public data class CalloutShapeProperties internal constructor(
```
Updated to: 
```kotlin
public fun Callout(
    geoElement: GeoElement,
    modifier: Modifier = Modifier,
    tapLocation: Point? = null,
    colorScheme: CalloutColors = CalloutDefaults.colors(),
    shape: CalloutShapes = CalloutDefaults.shapes(),
    content: @Composable BoxScope.() -> Unit
) {
```

shubham7109 commented:
@sorenoid Thanks for the feedback, I've updated the PR with the chang requests and the description test code. Ready for your review.

sorenoid reviewed (approved):
👍

kaushikrw commented on toolkit/geoview-compose/src/main/java/com/arcgismaps/toolkit/geoviewcompose/theme/CalloutTheme.kt:
```diff
@@ -0,0 +1,162 @@
+}
+
+@Immutable
+public data class CalloutShapes internal constructor(
```
Missing doc for parameters. The doc should ideally define what each property corresponds to in the UI.

kaushikrw commented on toolkit/geoview-compose/src/main/java/com/arcgismaps/toolkit/geoviewcompose/theme/CalloutTheme.kt:
```diff
@@ -0,0 +1,162 @@
+
+
+@Immutable
+public data class CalloutColors internal constructor(
```
Missing doc for parameters.

kaushikrw commented on toolkit/geoview-compose/src/main/java/com/arcgismaps/toolkit/geoviewcompose/theme/CalloutTheme.kt:
```diff
@@ -0,0 +1,162 @@
+     * @since 200.5.0
+     */
+    @Composable
+    public fun colors(
```
Missing doc for parameters.

kaushikrw commented on toolkit/geoview-compose/src/main/java/com/arcgismaps/toolkit/geoviewcompose/theme/CalloutTheme.kt:
```diff
@@ -0,0 +1,162 @@
+     * @since 200.5.0
+     */
+    @Composable
+    public fun shapes(
```
Missing doc for parameters.

kaushikrw commented on toolkit/geoview-compose/src/main/java/com/arcgismaps/toolkit/geoviewcompose/GeoViewScope.kt:
```diff
@@ -239,6 +253,8 @@ public sealed class GeoViewScope protected constructor(private val geoView: GeoV
         offset: Offset,
         rotateOffsetWithGeoView: Boolean,
+        colorScheme: CalloutColors,
+        shape: CalloutShapes,
```
```suggestion
        shapes: CalloutShapes,
```

kaushikrw replied:
Missed this..

kaushikrw commented on toolkit/geoview-compose/src/main/java/com/arcgismaps/toolkit/geoviewcompose/GeoViewScope.kt:
```diff
@@ -116,6 +120,8 @@ public sealed class GeoViewScope protected constructor(private val geoView: GeoV
         offset: Offset = Offset.Zero,
         rotateOffsetWithGeoView: Boolean = false,
+        colorScheme: CalloutColors = CalloutDefaults.colors(),
+        shape: CalloutShapes = CalloutDefaults.shapes(),
```
```suggestion
        shapes: CalloutShapes = CalloutDefaults.shapes(),
```

kaushikrw commented on toolkit/geoview-compose/src/main/java/com/arcgismaps/toolkit/geoviewcompose/GeoViewScope.kt:
```diff
@@ -153,6 +161,8 @@ public sealed class GeoViewScope protected constructor(private val geoView: GeoV
         modifier: Modifier = Modifier,
         tapLocation: Point? = null,
+        colorScheme: CalloutColors = CalloutDefaults.colors(),
+        shape: CalloutShapes = CalloutDefaults.shapes(),
```
```suggestion
        shapes: CalloutShapes = CalloutDefaults.shapes(),
```

kaushikrw commented on toolkit/geoview-compose/src/main/java/com/arcgismaps/toolkit/geoviewcompose/GeoViewScope.kt:
```diff
@@ -202,6 +212,8 @@ public sealed class GeoViewScope protected constructor(private val geoView: GeoV
         modifier: Modifier = Modifier,
         tapLocation: Point? = null,
+        colorScheme: CalloutColors,
+        shape: CalloutShapes,
```
```suggestion
        shapes: CalloutShapes,
```

kaushikrw commented on toolkit/geoview-compose/src/main/java/com/arcgismaps/toolkit/geoviewcompose/GeoViewScope.kt:
```diff
@@ -278,22 +294,25 @@ public sealed class GeoViewScope protected constructor(private val geoView: GeoV
-                )
-                {
-                    content.invoke(this)
+                CalloutTheme {
+                    with(LocalDensity.current) {
+                        Box(
+                            modifier = modifier
+                                .drawCalloutContainer(
+                                    cornerRadius = shape.cornerRadius.toPx(),
+                                    strokeBorderWidth = shape.borderWidth.toPx(),
+                                    strokeColor = colorScheme.borderColor,
+                                    backgroundColor = colorScheme.backgroundColor,
+                                    calloutContentPadding = shape.calloutContentPadding,
+                                    leaderWidth = shape.leaderSize.width.toPx(),
+                                    leaderHeight = shape.leaderSize.height.toPx(),
+                                    minSize = shape.minSize
+                                )
+                                .animateContentSize()
+                        ) {
+                            content.invoke(this)
+                        }
+                    }
```
If the `drawCalloutContainer` already accepts all these parameters then I don't see a need for the `CalloutTheme`. It is not being used at all. A theme and the use of composition local makes sense when the hierarchy is much more complex and deep. 

In this case, I think the `CalloutTheme` and their CompositionLocal's can be completely eliminated and you could simply pass the `CalloutShapes` and `CalloutColors` down to this modifier.

shubham7109 replied:
Thanks @kaushikrw, this helps clarify the usage of the theme.

kaushikrw reviewed (changes requested):
I think the `CalloutTheme` is not needed. I've added some context below..

shubham7109 commented:
@kaushikrw Ready for your re-review. I added the missing doc, removed the use of CalloutTheme & composition local, and changed the shape param to "shapes".

shubham7109 commented:
Ah sorry, thanks for pointing it out. Ready for your review @kaushikrw

kaushikrw reviewed (approved):
LGTM!
